File: events/2015/18/solution.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"time"
	"yod_aoc/utils/go/aoc_utils"
)

func countLight(state [][]bool, y int, x int, count int) int {
	if state[y][x] {
		return count + 1
	}
	return count
}

func countNeighbours(state [][]bool) [][]int {
	gridSize := len(state)
	count := make([][]int, gridSize)

	// Build an empty count from state
	for y := 0; y < gridSize; y++ {
		count[y] = make([]int, gridSize)
	}

	// Work through every position
	for y := 0; y < gridSize; y++ {
		for x := 0; x < gridSize; x++ {
			countForLight := 0

			// Check the row Above
			if y != 0 {
				if x != 0 {
					countForLight = countLight(state, y-1, x-1, countForLight)
				}
				countForLight = countLight(state, y-1, x, countForLight)
				if x != gridSize-1 {
					countForLight = countLight(state, y-1, x+1, countForLight)
				}
			}

			// Check the current Row
			if x != 0 {
				countForLight = countLight(state, y, x-1, countForLight)
			}
			if x != gridSize-1 {
				countForLight = countLight(state, y, x+1, countForLight)
			}

			// Count the row below
			if y+1 != gridSize {
				if x != 0 {
					countForLight = countLight(state, y+1, x-1, countForLight)
				}
				countForLight = countLight(state, y+1, x, countForLight)
				if x != gridSize-1 {
					countForLight = countLight(state, y+1, x+1, countForLight)
				}
			}

			// Update the main count
			count[y][x] = countForLight
		}
	}

	return count
}

func forceCornersOn(state [][]bool, gridSize int) [][]bool {
	gridSize -= 1
	state[0][0] = true
	state[0][gridSize] = true
	state[gridSize][0] = true
	state[gridSize][gridSize] = true
	return state
}

// ----------------------------------------------------------------------------------------------------
// | Part 1
// ----------------------------------------------------------------------------------------------------
func part1(input []string) int {
	var state [][]bool
	gridSizeY := 0
	gridSizeX := 0
	iterations := 100

	// Read through the input and build the state
	for i := 0; i < len(input); i++ {
		line := input[i]

		if len(line) == 0 {
			// This is a blank line, ignore it
			continue
		}

		// Increment GridSizeY
		gridSizeY += 1

		// Set the length of the line
		if gridSizeX == 0 {
			gridSizeX = len(line)
		} else if gridSizeX != len(line) {
			panic("Malformed input")
		}

		state = append(state, make([]bool, len(line)))
		for li := 0; li < len(line); li++ {
			if line[li] == '#' {
				state[i][li] = true
			} else {
				state[i][li] = false
			}
		}
	}

	// Run the iterations
	for it := 0; it < iterations; it++ {
		// Get the count for the current state
		count := countNeighbours(state)

		for y := 0; y < gridSizeY; y++ {
			for x := 0; x < gridSizeX; x++ {
				countForLight := count[y][x]

				if state[y][x] {
					if countForLight == 2 || countForLight == 3 {
						state[y][x] = true
					} else {
						state[y][x] = false
					}
				} else {
					if countForLight == 3 {
						state[y][x] = true
					} else {
						state[y][x] = false
					}
				}
			}
		}
	}

	// Count how many are on in the final state
	countOn := 0

	for y := 0; y < gridSizeY; y++ {
		for x := 0; x < gridSizeX; x++ {
			if state[y][x] {
				countOn += 1
			}
		}
	}

	return countOn
}

// ----------------------------------------------------------------------------------------------------
// | Part 2
// ----------------------------------------------------------------------------------------------------
func part2(input []string) int {
	var state [][]bool
	gridSizeY := 0
	gridSizeX := 0
	iterations := 100

	// Read through the input and build the state
	for i := 0; i < len(input); i++ {
		line := input[i]

		if len(line) == 0 {
			// This is a blank line, ignore it
			continue
		}

		// Increment GridSizeY
		gridSizeY += 1

		// Set the length of the line
		if gridSizeX == 0 {
			gridSizeX = len(line)
		} else if gridSizeX != len(line) {
			panic("Malformed input")
		}

		state = append(state, make([]bool, len(line)))
		for li := 0; li < len(line); li++ {
			if line[li] == '#' {
				state[i][li] = true
			} else {
				state[i][li] = false
			}
		}
	}

	// Force the corners on
	state = forceCornersOn(state, gridSizeX)

	// Run the iterations
	for it := 0; it < iterations; it++ {
		// Get the count for the current state
		count := countNeighbours(state)

		for y := 0; y < gridSizeY; y++ {
			for x := 0; x < gridSizeX; x++ {
				countForLight := count[y][x]

				if state[y][x] {
					if countForLight == 2 || countForLight == 3 {
						state[y][x] = true
					} else {
						state[y][x] = false
					}
				} else {
					if countForLight == 3 {
						state[y][x] = true
					} else {
						state[y][x] = false
					}
				}
			}
		}

		// Force the corners on
		state = forceCornersOn(state, gridSizeX)
	}

	// Count how many are on in the final state
	countOn := 0

	for y := 0; y < gridSizeY; y++ {
		for x := 0; x < gridSizeX; x++ {
			if state[y][x] {
				countOn += 1
			}
		}
	}

	return countOn
}

// ----------------------------------------------------------------------------------------------------
// | Main Function
// ----------------------------------------------------------------------------------------------------
func main() {
	// Read the arguments
	args := os.Args[1:]
	var fileName string
	if len(args) > 0 {
		fileName = args[0]
	} else {
		fileName = "input.txt"
	}
	// Start the timer!
	startTime := time.Now()

	// Read the input to an array of strings
	path := aoc_utils.GetFilePath(2015, 18, fileName)
	content, err := aoc_utils.ReadInputAsArray(path)
	if err != nil {
		fmt.Println("Error reading input:", err)
		return
	}

	// Run the Parts
	p1Result := aoc_utils.MeasurePerformance(func() interface{} {
		return part1(content)
	})

	p2Result := aoc_utils.MeasurePerformance(func() interface{} {
		return part2(content)
	})

	// End the Timer!
	endTime := time.Now()
	duration := endTime.Sub(startTime)

	resultsForDay := aoc_utils.AOCDayResults{
		Year:     2015,
		Day:      18,
		Part1:    p1Result,
		Part2:    p2Result,
		Duration: duration.Nanoseconds(),
		Timestamp: aoc_utils.AOCTimestamp{
			Start: startTime.Format(time.RFC3339),
			End:   endTime.Format(time.RFC3339),
		},
	}
	jsonResults, _ := json.Marshal(resultsForDay)
	fmt.Println(string(jsonResults))
}

File: events/2015/18/solution.rs
use aoc_utils::read_input_as_array;
use aoc_utils::measure_performance;
use aoc_utils::AOCDayResults;
use aoc_utils::AOCTimestamp;
use chrono::{DateTime, Utc};
use std::collections::HashMap;

fn count_light(state: &HashMap<i32, HashMap<i32, bool>>, y: i32, x: i32, count: i32) -> i32 {
  if *state.get(&y).and_then(|row| row.get(&x)).unwrap_or(&false) {
    return count + 1;
  }
  count
}

fn count_neighbours(state: &HashMap<i32, HashMap<i32, bool>>) -> HashMap<i32, HashMap<i32, i32>> {
  let mut count: HashMap<i32, HashMap<i32, i32>> = HashMap::new();
  let grid_size = state.len() as i32;

  // Build an empty count from state
  for y in 0..grid_size {
    count.insert(y, HashMap::new());
    for x in 0..grid_size {
      count.get_mut(&y).unwrap().insert(x, 0);
    }
  }

  // Work through every position
  for y in 0..grid_size {
    for x in 0..grid_size {
      let mut count_for_light = 0;

      // Check the row Above
      if y != 0 {
        if x != 0 {
          count_for_light = count_light(state, y - 1, x - 1, count_for_light);
        }
        count_for_light = count_light(state, y - 1, x, count_for_light);
        if x != grid_size - 1 {
          count_for_light = count_light(state, y - 1, x + 1, count_for_light);
        }
      }

      // Check the current Row
      if x != 0 {
        count_for_light = count_light(state, y, x - 1, count_for_light);
      }
      if x != grid_size - 1 {
        count_for_light = count_light(state, y, x + 1, count_for_light);
      }

      // Count the row below
      if y + 1 != grid_size {
        if x != 0 {
          count_for_light = count_light(state, y + 1, x - 1, count_for_light);
        }
        count_for_light = count_light(state, y + 1, x, count_for_light);
        if x != grid_size - 1 {
          count_for_light = count_light(state, y + 1, x + 1, count_for_light);
        }
      }

      // Update the main count
      count.get_mut(&y).unwrap().insert(x, count_for_light);
    }
  }

  count
}

fn force_corners_on(state: &mut HashMap<i32, HashMap<i32, bool>>, grid_size: i32) {
  let grid_size = grid_size - 1;
  state.get_mut(&0).unwrap().insert(0, true);
  state.get_mut(&0).unwrap().insert(grid_size, true);
  state.get_mut(&grid_size).unwrap().insert(0, true);
  state.get_mut(&grid_size).unwrap().insert(grid_size, true);
}

// ----------------------------------------------------------------------------------------------------
// | Part 1
// ----------------------------------------------------------------------------------------------------
fn part1(input: &Vec<String>) -> i32 {
  let mut state: HashMap<i32, HashMap<i32, bool>> = HashMap::new();
  let mut grid_size_y = 0;
  let mut grid_size_x = 0;
  let iterations = 100;

  // Read through the input and build the state
  for (i, line) in input.iter().enumerate() {
    if line.is_empty() {
      // This is a blank line, ignore it
      continue;
    }

    // Increment GridSizeY
    grid_size_y += 1;

    // Set the length of the line
    if grid_size_x == 0 {
        grid_size_x = line.len() as i32;
    } else if grid_size_x != line.len() as i32 {
      eprintln!("Malformed input");
      return 0;
    }

    let i = i as i32;
    state.insert(i, HashMap::new());
    for (li, ch) in line.chars().enumerate() {
      let value = ch == '#';
      state.get_mut(&i).unwrap().insert(li as i32, value);
    }
  }

  // Run the iterations
  for _it in 0..iterations {
    // Get the count for the current state
    let count = count_neighbours(&state);

    for y in 0..grid_size_y {
      for x in 0..grid_size_x {
        let count_for_light = *count.get(&y).and_then(|row| row.get(&x)).unwrap();

        let new_value = if *state.get(&y).and_then(|row| row.get(&x)).unwrap() {
          count_for_light == 2 || count_for_light == 3
        } else {
          count_for_light == 3
        };
        state.get_mut(&y).unwrap().insert(x, new_value);
      }
    }
  }

  // Count how many are on in the final state
  let mut count_on = 0;

  for y in 0..grid_size_y {
    for x in 0..grid_size_x {
      if *state.get(&y).and_then(|row| row.get(&x)).unwrap() {
        count_on += 1;
      }
    }
  }

  count_on
}

// ----------------------------------------------------------------------------------------------------
// | Part 2
// ----------------------------------------------------------------------------------------------------
fn part2(input: &Vec<String>) -> i32 {
  let mut state: HashMap<i32, HashMap<i32, bool>> = HashMap::new();
  let mut grid_size_y = 0;
  let mut grid_size_x = 0;
  let iterations = 100;

  // Read through the input and build the state
  for (i, line) in input.iter().enumerate() {
    if line.is_empty() {
      // This is a blank line, ignore it
      continue;
    }

    // Increment GridSizeY
    grid_size_y += 1;

    // Set the length of the line
    if grid_size_x == 0 {
      grid_size_x = line.len() as i32;
    } else if grid_size_x != line.len() as i32 {
      eprintln!("Malformed input");
      return 0;
    }

    let i = i as i32;
    state.insert(i, HashMap::new());
    for (li, ch) in line.chars().enumerate() {
      let value = ch == '#';
      state.get_mut(&i).unwrap().insert(li as i32, value);
    }
  }

  // Force the corners on
  force_corners_on(&mut state, grid_size_x);

  // Run the iterations
  for _it in 0..iterations {
    // Get the count for the current state
    let count = count_neighbours(&state);

    for y in 0..grid_size_y {
      for x in 0..grid_size_x {
        let count_for_light = *count.get(&y).and_then(|row| row.get(&x)).unwrap();

        let new_value = if *state.get(&y).and_then(|row| row.get(&x)).unwrap() {
          count_for_light == 2 || count_for_light == 3
        } else {
          count_for_light == 3
        };
        state.get_mut(&y).unwrap().insert(x, new_value);
      }
    }

    // Force the corners on
    force_corners_on(&mut state, grid_size_x);
  }

  // Count how many are on in the final state
  let mut count_on = 0;

  for y in 0..grid_size_y {
    for x in 0..grid_size_x {
      if *state.get(&y).and_then(|row| row.get(&x)).unwrap() {
        count_on += 1;
      }
    }
  }

  count_on
}

// ----------------------------------------------------------------------------------------------------
// | Main Function
// ----------------------------------------------------------------------------------------------------
fn main() {
  // Read the arguments
  let args = std::env::args().collect::<Vec<String>>();
  let default_file_name = String::from("input.txt");
  let file_name = args.get(1).unwrap_or(&default_file_name);

  // Start the timer
  let start_timestamp = std::time::SystemTime::now();
  let start_time = std::time::Instant::now();
  
  // Read the input
  let input = read_input_as_array(file_name).unwrap();

  // Run the Parts
  let p1_result = measure_performance(part1, &input);
  let p2_result = measure_performance(part2, &input);

  // Stop the timer
  let end_time = std::time::Instant::now();
  let end_timestamp = std::time::SystemTime::now();
  let duration = end_time.duration_since(start_time).as_nanos();

  let results = AOCDayResults {
    year: 2015,
    day: 18,
    part1: p1_result,
    part2: p2_result,
    duration: duration,
    timestamp: AOCTimestamp {
      start: DateTime::<Utc>::from(start_timestamp).to_rfc3339(),
      end: DateTime::<Utc>::from(end_timestamp).to_rfc3339(),
    },
  };

  let json_results = serde_json::to_string(&results).unwrap();
  println!("{}", json_results);
}